core/cog-gamepad.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * cog-gamepad.c
+ * Copyright (C) 2022 Igalia S.L.
+ *
+ * Distributed under terms of the MIT license.
+ */
+
+#include "cog.h"
+#include <wpe/wpe.h>
+
+#if COG_ENABLE_GAMEPAD_MANETTE
+extern const struct wpe_gamepad_provider_interface manette_provider_interface;
+extern const struct wpe_gamepad_interface          manette_device_interface;
+#endif
+
+static const struct CogGamepadBackend {
+    const char                                  *name;
+    const struct wpe_gamepad_provider_interface *provider;
+    const struct wpe_gamepad_interface          *device;
+} s_gamepad_backends[] = {
+#if COG_ENABLE_GAMEPAD_MANETTE
+    {
+        "manette",
+        &manette_provider_interface,
+        &manette_device_interface,
+    },
+#endif
+    {
+        "disabled",
+        NULL,
+        NULL,
+    },
+};
+
+/* selected gamepad backend name */
+static const char *backend;
+/* selected gamepad backend interfaces */
+static const struct CogGamepadBackend *selected;
+/* selected gamepad backend provider interfaces */
+static struct wpe_gamepad_provider_interface provider_interface;
+
+void
+cog_gamepad_set_backend(const char *name)
+{
+    if (backend)
+        return;
+
+    if (!name) {
+        backend = s_gamepad_backends[0].name;
+        goto bail;
+    }
+
+    for (int i = 0; i < G_N_ELEMENTS(s_gamepad_backends); i++)
+        if (g_strcmp0(name, s_gamepad_backends[i].name) == 0)
+            backend = s_gamepad_backends[i].name;
+
+bail:
+    g_debug("gamepad backend: %s", backend);
+}
+
+void
+cog_gamepad_setup(GamepadProviderGetViewBackend *gamepad_get_view)
+{
+    if (selected)
+        return;
+
+    if (!backend)
+        return;
```

Comment:
Doing as suggested, the `s_backend` pointer will always be valid, and we only need to check that this function is called only once.

```suggestion
    static bool initialized = false;
    g_return_if_fail(!initialized);

    initialized = true;
```